File: src/main/java/org/sunflow/math/Point3.java
package org.sunflow.math;

public final class Point3 {
    public float x, y, z;

    public Point3() {
    }

    public Point3(float x, float y, float z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Point3(Point3 p) {
        x = p.x;
        y = p.y;
        z = p.z;
    }

    public float get(int i) {
        switch (i) {
            case 0:
                return x;
            case 1:
                return y;
            default:
                return z;
        }
    }

    public final float distanceTo(Point3 p) {
        float dx = x - p.x;
        float dy = y - p.y;
        float dz = z - p.z;
        return (float) Math.sqrt((dx * dx) + (dy * dy) + (dz * dz));
    }

    public final float distanceTo(float px, float py, float pz) {
        float dx = x - px;
        float dy = y - py;
        float dz = z - pz;
        return (float) Math.sqrt((dx * dx) + (dy * dy) + (dz * dz));
    }

    public final float distanceToSquared(Point3 p) {
        float dx = x - p.x;
        float dy = y - p.y;
        float dz = z - p.z;
        return (dx * dx) + (dy * dy) + (dz * dz);
    }

    public final float distanceToSquared(float px, float py, float pz) {
        float dx = x - px;
        float dy = y - py;
        float dz = z - pz;
        return (dx * dx) + (dy * dy) + (dz * dz);
    }

    public final Point3 set(float x, float y, float z) {
        this.x = x;
        this.y = y;
        this.z = z;
        return this;
    }

    public final Point3 set(Point3 p) {
        x = p.x;
        y = p.y;
        z = p.z;
        return this;
    }

    public static final Point3 add(Point3 p, Vector3 v, Point3 dest) {
        dest.x = p.x + v.x();
        dest.y = p.y + v.y();
        dest.z = p.z + v.z();
        return dest;
    }

    public final Vector3 sub(Point3 p) {
        return Vector3J.create(x - p.x, y - p.y, z - p.z);
    }

    public static final Point3 mid(Point3 p1, Point3 p2, Point3 dest) {
        dest.x = 0.5f * (p1.x + p2.x);
        dest.y = 0.5f * (p1.y + p2.y);
        dest.z = 0.5f * (p1.z + p2.z);
        return dest;
    }

    public static final Point3 blend(Point3 p0, Point3 p1, float blend, Point3 dest) {
        dest.x = (1 - blend) * p0.x + blend * p1.x;
        dest.y = (1 - blend) * p0.y + blend * p1.y;
        dest.z = (1 - blend) * p0.z + blend * p1.z;
        return dest;
    }

    public static final Vector3 normal(Point3 p0, Point3 p1, Point3 p2) {
        Vector3 edge1 = p1.sub(p0);
        Vector3 edge2 = p2.sub(p0);
        Vector3 c = Vector3J.cross(edge1, edge2);
        /** TODO: Ideally, we shouldn't be trying to compute the normal for
         *  collinear points (it's impossible).
        if (c.length() <= 0) {
            n = Vector3J.create(1, 0, 0);
            UI.printWarning(Module.GEOM, "Was asked for the normal to three collinear points!");
        } else {
            
        }
        */
        Vector3 n = Vector3J.normalize(c); 
        return n; 
    }

    @Override
    public final String toString() {
        return String.format("(%.2f, %.2f, %.2f)", x, y, z);
    }
}

File: src/main/scala/org/sunflow/math/Point3.scala
package org.sunflow.math

trait Point3 {
  def x: Float
  def y: Float
  def z: Float
  
  def +(v: Vector3)(implicit b: Point3Builder): Point3 =
    b(x + v.x, y + v.y, z + v.z)
  def -(v: Vector3)(implicit b: Point3Builder): Point3 =
    b(x - v.x, y - v.y, z - v.z)
  def -(p: Point3)(implicit b: Vector3Builder): Vector3 =
    b(x - p.x, y - p.y, z - p.z)
  def mid(p: Point3)(implicit b: Point3Builder): Point3 =
    this + (p - this) * 0.5f
  def lerpTo(p: Point3, pAmount: Float)(implicit b: Point3Builder): Point3 =
    this + (p - this) * pAmount
  
  def distanceTo(p: Point3): Float = (this - p).length
  def distanceToSquared(p: Point3): Float = (this - p).lengthSquared
}

trait Point3Builder extends ((Float, Float, Float) => Point3)

object Point3 {
  private implicit val normal3Builder: Normal3Builder = 
    Normal3.SimpleNormal3Builder
  def apply(x: Float, y: Float, z: Float)(implicit b: Point3Builder): Point3 =
    b(x, y, z)
  def normal(p0: Point3, p1: Point3, p2: Point3): Normal3 =
    ((p1 - p0) cross (p2 - p0)).normalize
  private [Point3] final case class SimplePoint3(
      x: Float, y: Float, z: Float) extends Point3
  private [Point3] sealed class SimplePoint3Builder extends Point3Builder {
    def apply(x: Float, y: Float, z: Float): Point3 = SimplePoint3(x, y, z)
  }
  implicit object SimplePoint3Builder extends SimplePoint3Builder
  val Zero = SimplePoint3(0, 0, 0)
}

// Static call interface for Java.
object Point3J {
  def create(x: Float, y: Float, z: Float): Point3 = Point3(x, y, z)
  def add(p: Point3, v: Vector3): Point3 = p + v
  def sub(p: Point3, v: Vector3): Point3 = p - v
  def sub(p0: Point3, p1: Point3): Vector3 = p0 - p1
  def mid(p0: Point3, p1: Point3): Point3 = p0 mid p1
  def lerpTo(p0: Point3, p1: Point3, p1Amount: Float): Point3 = 
    p0.lerpTo(p1, p1Amount)
  def normal(p0: Point3, p1: Point3, p2: Point3): Normal3 =
    Point3.normal(p0, p1, p2)
  def zero(): Point3 = Point3.Zero
}
